src/compiler/checker.ts

```diff
@@ -14272,20 +14272,38 @@ namespace ts {
         function discriminateTypeByDiscriminableItems(target: UnionType, discriminators: [() => Type, __String][], related: (source: Type, target: Type) => boolean | Ternary): Type | undefined;
         function discriminateTypeByDiscriminableItems(target: UnionType, discriminators: [() => Type, __String][], related: (source: Type, target: Type) => boolean | Ternary, defaultValue: Type): Type;
         function discriminateTypeByDiscriminableItems(target: UnionType, discriminators: [() => Type, __String][], related: (source: Type, target: Type) => boolean | Ternary, defaultValue?: Type) {
-            let match: Type | undefined;
+            let rounds: number | undefined;
+            let candidates: Map<number> | undefined;
             for (const [getDiscriminatingType, propertyName] of discriminators) {
+                if (rounds !== undefined) {
+                    rounds++;
+                }
                 for (const type of target.types) {
                     const targetType = getTypeOfPropertyOfType(type, propertyName);
                     if (targetType && related(getDiscriminatingType(), targetType)) {
-                        if (match) {
-                            if (type === match) continue; // Finding multiple fields which discriminate to the same type is fine
-                            return defaultValue;
+                        if (rounds === undefined) {
+                            rounds = 1;
+                            candidates = createMap();
+                            candidates.set("" + type.id, 1);
+                            continue;
                         }
-                        match = type;
+                        const strId = "" + type.id;
+                        candidates!.set(strId, (candidates!.get(strId) || 0) + 1);
+                    }
+                }
+            }
+            if (candidates) {
+                let matchID: string | false | undefined;
+                candidates.forEach((value, id) => {

```

Comment:
A comment explaining what criteria this is filtering/selecting on would be useful. I think it's that it returns the first type which matches all discriminant properties (out of multiple which may match), but some text expressing that would be nice.

Reply:
Will do. If there are multiple valid matches it fails rather than picking the first. I did this because to avoid situations where the order of the union influenced whether you got an error, sometimes incorrectly. (I think it was a case where some constituents were technically subtypes of another, but the union was unreduced).

Reply:
@weswigham Done. In the case where there are multiple candidates would could return a new union of just those candidates, but I'm not sure of the error message and perf consequences---if you want me to experiment with that approach let me know.